drivers/spi/spi_gd32.c

```diff
@@ -195,6 +233,146 @@ static int spi_gd32_frame_exchange(const struct device *dev)
+	dma_cfg->head_block = block_cfg;
+	dma_cfg->dma_slot = cfg->dma[dir].slot;
+	dma_cfg->channel_priority =
+		GD32_DMA_CONFIG_PRIORITY(cfg->dma[dir].config);
+	dma_cfg->channel_direction =
+		dir == TX ? MEMORY_TO_PERIPHERAL : PERIPHERAL_TO_MEMORY;
+
+	if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
+		dma_cfg->source_data_size = 1;
+		dma_cfg->dest_data_size = 1;
+	} else {
+		dma_cfg->source_data_size = 2;
+		dma_cfg->dest_data_size = 2;
+	}
+
+	block_cfg->block_size = spi_context_max_continuous_chunk(&data->ctx);
+
+	if (dir == TX) {
+		block_cfg->dest_address = datareg;
+		block_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			block_cfg->source_address = (uint32_t)data->ctx.tx_buf;
+			block_cfg->source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			block_cfg->source_address = (uint32_t)&dummy_tx;
+			block_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+
+	if (dir == RX) {
+		block_cfg->source_address = datareg;
+		block_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+
+		if (spi_context_rx_buf_on(&data->ctx)) {
+			block_cfg->dest_address = (uint32_t)data->ctx.rx_buf;
+			block_cfg->dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			block_cfg->dest_address = (uint32_t)&dummy_rx;
+			block_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+
+	ret = dma_config(dma->dev, dma->channel, dma_cfg);
+	if (ret < 0) {
+		LOG_ERR("dma_config %p failed %d\n", dma->dev, ret);
+		return ret;
+	}
+
+	ret = dma_start(dma->dev, dma->channel);
+	if (ret < 0) {
+		LOG_ERR("dma_start %p failed %d\n", dma->dev, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int spi_gd32_start_dma_transceive(const struct device *dev)
+{
+	const struct spi_gd32_config *cfg = dev->config;
+	struct spi_gd32_data *data = dev->data;
+	const size_t chunk_len = spi_context_max_continuous_chunk(&data->ctx);
+	struct dma_status stat;
+	int ret = 0;
+
+	if (!cfg->dma[TX].dev && spi_context_tx_on(&data->ctx)) {
+		LOG_ERR("DMA for TX not configured.");
+		return -EINVAL;
+	}
+
+	if (!cfg->dma[RX].dev && spi_context_rx_on(&data->ctx)) {
+		LOG_ERR("DMA for RX not configured.");
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(cfg->dma); i++) {
+		if (!cfg->dma[i].dev) {
+			continue;
+		}
+
+		dma_get_status(cfg->dma[i].dev, cfg->dma[i].channel, &stat);
+		if ((chunk_len != data->dma[i].count) && !stat.busy) {
+			ret = spi_gd32_dma_setup(dev, i);
+			if (ret < 0) {
+				goto on_error;
+			}
+		}
+	}
+
+	SPI_CTL1(cfg->reg) |= (SPI_CTL1_DMATEN | SPI_CTL1_DMAREN);
+
+on_error:
+	if (ret < 0) {
+		for (size_t i = 0; i < ARRAY_SIZE(cfg->dma); i++) {
+			if (!cfg->dma[i].dev) {
+				continue;
+			}
+			dma_stop(cfg->dma[i].dev, cfg->dma[i].channel);
+		}
+	}
+	return ret;
+}
+#endif
+
+static inline bool spi_gd32_dma_enabled(const struct device *dev)
+{
+#ifdef CONFIG_SPI_GD32_DMA
+	const struct spi_gd32_config *cfg = dev->config;
+
+	if (cfg->dma[TX].dev || cfg->dma[RX].dev) {
```

Comment:
Shall we use `&&` in here? TX and RX dma can not work individually.

Reply:
Change to require setting both TX and RX DMA channels.
Change to && here.